### ISSUE TITLE
Repo: Rule [options] parameter should be non-nullable if defaultOptions exists

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#5439]
### Suggestion

Right now, if a rule is created through `util.createRule` with `defaultOptions`, the `create` function's second parameter is an array whose first element is still `Options | undefined`. That means folks have to `!`. For example, in #5327:

https://github.com/typescript-eslint/typescript-eslint/blob/7f2661fd0a8f73064a68b5c6872bbc8f0a3c0594/packages/eslint-plugin/src/rules/lines-around-comment.ts#L136-L137

Unless I've grossly misinterpreted the rule options, those `!` should be unnecessary, right?

**Comment by JoshuaKGoldberg** [comment]
We've deleted `lines-around-comment` in the `v8` branch. A new still-in-the-repo example is `consistent-type-imports`:

https://github.com/typescript-eslint/typescript-eslint/blob/e408b93e48e5199d83a8d99d1e27126d2dd8bc8f/packages/eslint-plugin/src/rules/consistent-type-imports.ts#L89-L98

**Comment by JoshuaKGoldberg** [comment]
#9025 touches a lot of the same interfaces, so marking as blocked on that one till it's merged.

**Comment by JoshuaKGoldberg** [comment]
I took a run at this today. It was, as expected, very tricky to try to get to work. Removing from the `v8` milestone as there are two upstream ecosystem work items that could land first:

* https://github.com/eslint/eslint/pull/17656: which may totally change how we'd want to handle default options anyway
* https://github.com/Microsoft/TypeScript/pull/26349, per the note in `RuleCreator`

On that latter list item, `RuleCreator`'s `createNamedRule` is pretty complex right now:

https://github.com/typescript-eslint/typescript-eslint/blob/b1c92d4a1a0448ca07889fd180dedcc6c85c63c5/packages/utils/src/eslint-utils/RuleCreator.ts#L59-L64

Having to add a potential third type parameter for `OptionDefaults` is ... even more complexity. Especially since we currently sometimes need to be explicit about types in `createRule` calls:

https://github.com/typescript-eslint/typescript-eslint/blob/04990d545fc119329551ae3a55d79dfc0c7bf147/packages/eslint-plugin/src/rules/no-floating-promises.ts#L52

Marking as blocked on external items. 😞

**Comment by JoshuaKGoldberg** [comment]
Since the ESLint RFC, I'll go ahead and close this out. We can always re-file a new issue if we see a similar path forward with `meta.defaultOptions`. Cheers!
